python/cudf/cudf/core/column/column.py

```diff
@@ -120,6 +120,13 @@ def mask_array_view(self):
     def __len__(self):
         return self.size
 
+    def __iter__(self):
+        raise TypeError(
+            "Creation of an Iterator over a cudf Column is not allowed, "
+            "To create an iterator, explicitly convert to any of the objects "
+            "supporting iteration using .to_pandas(), .to_arrow()"
```

Comment:
Should wrap API names in tickmarks to make it easier to read and separate APIs from punctuation.

Reply:
I think just not implementing `__iter__` if we don't want to support iteration is enough, but I see the virtues of a descriptive error message.

I think it's better not to make any assumptions in our error message about how/why the `__iter__` method is being called:

```python
f"{self.__class__.__name__} object is not iterable. Consider using `.to_arrow()`, `.to_pandas()` or `.values_host` if you wish to iterate over the values."
```

Ditto below.

Reply:
Actually, I'm feeling a bit more strongly that we shouldn't implement `__iter__` at all if we don't want our classes to be iterable.

```python
In [43]: class Foo(): 
    ...:     def __iter__(self): 
    ...:         raise TypeError() 
    ...:                                                                                                                                                                          

In [44]: from typing import Iterable                                                                                                                                              

In [45]: isinstance(Foo(), Iterable)                                                                                                                                              
Out[45]: True
```

This feels super un-pythonic to me.

Reply:
I think we might need the `__iter__` if we want to show a descriptive error message providing an alternative when the user tries to iterate.

or else I think we might leave the users in a puzzled state as in why iteration is not supported and what the alternative is?

Reply:
My understanding is that if we don't implement `__iter__` it ends up falling back to `__getitem__` eventually and iterating via that.

Reply:
One more reason why we will need to add and throw in `__iter__` is that if there is no `__iter__`, and there is `__getitem__` implemented, the iteration will happen via a call to `__getitem__`.

Reply:
> Actually, I'm feeling a bit more strongly that we shouldn't implement `__iter__` at all if we don't want our classes to be iterable.
> 
> ```python
> In [43]: class Foo(): 
>     ...:     def __iter__(self): 
>     ...:         raise TypeError() 
>     ...:                                                                                                                                                                          
> 
> In [44]: from typing import Iterable                                                                                                                                              
> 
> In [45]: isinstance(Foo(), Iterable)                                                                                                                                              
> Out[45]: True
> ```
> 
> This feels super un-pythonic to me.

I think the correct way to check for an iterable is :

```python
def iterable(obj):
    try:
        iter(obj)
    except Exception:
        return False
    else:
        return True
```

source: https://stackoverflow.com/a/61139278

Reply:
Yeah I understand. On the one hand, we prevent implicit DtoH by having an `__iter__()` that raises. On the other hand, we break some pretty fundamental rules by implementing `__iter__()` for an object not meant to be iterable.

It's possible there's a solution here. Could you please give me a bit to try and find one before merging this?

Reply:
Sure Ashwin 👍

Reply:
In the Python docs as well: https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable

> The only reliable way to determine whether an object is iterable is to call iter(obj).

Reply:
Thanks Keith - OK I think I made a bigger deal about this than it is then :) no objections from me. I'd still suggest the error message though

Reply:
Moved the changed error message to a common utility, Could you both review if it looks good?

Reply:
LGTM, resolving this conversation